### DISCUSSION TITLE
Submodule handling after git locked down file protocol

### DISCUSSION THREAD

**Discussion by vbjay** [gitextensions/gitextensions#10426] [Ideas]
In reference to #10387  and #10389

There are a few ways of going about the fix of this breakage that I see.

1. Ask user if it is ok to do recursive and that they trust the submodule chain.  If so, pass down -c protocol.file.allow=always to commands which will allow ```submodule add``` and ```update  --init --recursive``` 
2. Forgo --recursive in submodule update and such and instead do some kind of traversal and update per level ourselves.  Do some kind of 
```
You are about to clone/update... <submodule source uri here >.  Do you allow this? 
```
dialog that then passes the required config only when user accepts.  

3. Alert user that file://protocol submodules are unsafe and point to documentation about this issue and cve and such when trying to add and if user allows then pass in config.  

I am looking for thoughts and what we think should happen in these cases.  We could be smart about it and look at protocol and if not file then just let our external exception stuff show through but if file do all or some of the suggestions.

**Comment by vbjay** [comment]
Thoughts on risk and functionality

1.  Risky.  Does prevent aggravating user with a zillion do you allow but could allow in exactly what the security is designed to prevent.
2. Less risky.  Will probably annoy the *#$&*&$ out of users if they have a chain of submodules when they are adding/updating submodules.
3. No risk because we are just not allowing file protocol submodules anymore. Submodules will have to be some other trusted protocol.  

We can combine 3 and other possibilities to help explain why we need to ask if doing 1 or 2 or other action.

**Reply by RussKie** [reply]
> 3\. No risk because we are just not allowing file protocol submodules anymore. Submodules will have to be some other trusted protocol.

The onus would be on users to enable the protocol manually, is this correct?

**Reply by vbjay** [reply]
Sort of. In my testing I proved that at least currently setting a file config doesn't work. It must be passed in to work. This means adding or updating submodules need the -c... config item. So basically Your Milage May Vary.  Can play with the work I did to vet that idea. 

Basically I don't think file protocol submodule add or update will work without us passing command line config. So 3 basically says yes this doesn't work because ... In some kind of ui and tell them to use a different protocol or whatnot. 

```patch

From bb70ff745664bc395d66ae7aeb1a1c3fe882a280 Mon Sep 17 00:00:00 2001
From: Jay Asbury <vbjay.net@gmail.com>
Date: Wed, 23 Nov 2022 02:40:00 -0500
Subject: [PATCH] tmp

---
 UnitTests/CommonTestUtils/GitModuleTestHelper.cs | 4 ++--
 UnitTests/GitCommands.Tests/GitModuleTests.cs    | 2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/UnitTests/CommonTestUtils/GitModuleTestHelper.cs b/UnitTests/CommonTestUtils/GitModuleTestHelper.cs
index e6c70a783..3952de730 100644
--- a/UnitTests/CommonTestUtils/GitModuleTestHelper.cs
+++ b/UnitTests/CommonTestUtils/GitModuleTestHelper.cs
@@ -158,7 +158,7 @@ public void AddSubmodule(GitModuleTestHelper subModuleHelper, string path)
             Assert.That(fileEnabled == "always");
 
             // Even though above is set, adding a file protocol submodule fails unless -c... is used for protocol.file.allow config.
-            IEnumerable<GitConfigItem> cfgs = GitCommandHelpers.GetAllowFileConfig();
+            IEnumerable<GitConfigItem> cfgs = null; // GitCommandHelpers.GetAllowFileConfig();
 
             var result = Module.GitExecutable.Execute(GitCommandHelpers.AddSubmoduleCmd(subModuleHelper.Module.WorkingDir.ToPosixPath(), path, null, true, cfgs));
             Debug.WriteLine(result.AllOutput);
@@ -172,7 +172,7 @@ public void AddSubmodule(GitModuleTestHelper subModuleHelper, string path)
         /// <returns>All submodule Modules</returns>
         public IEnumerable<GitModule> GetSubmodulesRecursive()
         {
-            IEnumerable<GitConfigItem> configs = GitCommandHelpers.GetAllowFileConfig();
+            IEnumerable<GitConfigItem> configs = null; // GitCommandHelpers.GetAllowFileConfig();
             ArgumentString args = GitCommandHelpers.SubmoduleUpdateCmd(name: null, configs: configs);
 
             Module.GitExecutable.Execute(args);
diff --git a/UnitTests/GitCommands.Tests/GitModuleTests.cs b/UnitTests/GitCommands.Tests/GitModuleTests.cs
index 8ad9cf434..677e377a5 100644
--- a/UnitTests/GitCommands.Tests/GitModuleTests.cs
+++ b/UnitTests/GitCommands.Tests/GitModuleTests.cs
@@ -711,7 +711,7 @@ public void GetSubmodulesLocalPaths()
 
                 // Init all modules of root
                 var root = moduleTestHelpers[0];
-                IEnumerable<GitConfigItem> cfgs = GitCommandHelpers.GetAllowFileConfig();
+                IEnumerable<GitConfigItem> cfgs = null; // GitCommandHelpers.GetAllowFileConfig();
 
                 root.Module.GitExecutable.Execute(GitCommandHelpers.SubmoduleUpdateCmd(name: null, cfgs));
 
-- 
2.38.1.windows.1


```


![image](https://user-images.githubusercontent.com/3629489/203499127-a0cd4149-f5e7-414a-9d6e-5577bde7ba64.png)





See

https://github.com/gitextensions/gitextensions/commit/37f4a7fefd04d3afb7ba85c5945ba0612cc3b6fc#diff-b95371a2b66ff68b214ce4a405b73c8ee2408c383932629e33a2ce5305a305eeR142

and

https://github.com/gitextensions/gitextensions/commit/37f4a7fefd04d3afb7ba85c5945ba0612cc3b6fc#diff-b95371a2b66ff68b214ce4a405b73c8ee2408c383932629e33a2ce5305a305eeR156-R163

and

https://github.com/gitextensions/gitextensions/commit/37f4a7fefd04d3afb7ba85c5945ba0612cc3b6fc#diff-b95371a2b66ff68b214ce4a405b73c8ee2408c383932629e33a2ce5305a305eeR174-R177

and

https://github.com/gitextensions/gitextensions/commit/37f4a7fefd04d3afb7ba85c5945ba0612cc3b6fc#diff-52f3e4e801e19209c2022f311ba449e1d498eaae107117a6d6155ebee8a67c55R711-R715

**Reply by vbjay** [reply]
So as you can see from above pictures and the fact I am verifying effective setting is set to always, you can see that -c... on command line is REQUIRED and it does not matter if user sets the config in a git config file.

As shown by https://github.com/gitextensions/gitextensions/blob/37f4a7fefd04d3afb7ba85c5945ba0612cc3b6fc/UnitTests/CommonTestUtils/GitModuleTestHelper.cs#L156-L163

Debug any one of those tests.  Look at the config file in the repo.  It is being set.  Passing same config to the configs argument in the command which adds the -c... to the command line solves the problem.  SO I am afraid add submodules will have to allow it.  I hope this is a git for windows bug.  I should be able to call  submodule add and it work.  We may need to look at passing a depth argument in to make sure it only adds the submodule we are adding. Will look into.  depth = shallow clone and not recursive mitigation.  So that won't work.
